categories.go

```
@@ -26,129 +28,163 @@ type Category struct {
-				includePolicyTemplates, err = strconv.ParseBool(v)
-				if err != nil {
-					badRequest(w, fmt.Sprintf("invalid 'include_policy_templates' query param: '%s'", v))
-					return
-				}
-			}
+		cacheHeaders(w, cacheTime)
+		jsonHeader(w)
+		fmt.Fprint(w, string(data))
+	}
+}
+
+type categoriesFilter struct {
+	Experimental           bool
+	KibanaVersion          *semver.Version
+	IncludePolicyTemplates bool
+}
+
+func newCategoriesFilterFromParams(r *http.Request) (categoriesFilter, error) {
+	var filter categoriesFilter
+
+	query := r.URL.Query()
+	if len(query) == 0 {
+		return filter, nil
+	}
+
+	var err error
+	if v := query.Get("kibana.version"); v != "" {
+		filter.KibanaVersion, err = semver.NewVersion(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid Kibana version '%s': %w", v, err)
 		}
+	}
 
-		packageList := map[string]util.Package{}
-		// Get unique list of newest packages
-		for _, p := range packages {
-			// Check if the package is compatible with Kibana version
-			if kibanaVersion != nil {
-				if valid := p.HasKibanaVersion(kibanaVersion); !valid {
-					continue
-				}
-			}
+	if v := query.Get("experimental"); v != "" {
+		filter.Experimental, err = strconv.ParseBool(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid 'experimental' query param: '%s'", v)
+		}
+	}
 
-			// Skip internal packages
-			if p.Internal {
-				continue
-			}
+	if v := query.Get("include_policy_templates"); v != "" {
+		filter.IncludePolicyTemplates, err = strconv.ParseBool(v)
+		if err != nil {
+			return filter, fmt.Errorf("invalid 'include_policy_templates' query param: '%s'", v)
+		}
+	}
 
-			// Skip experimental packages if flag is not specified
-			if p.Release == util.ReleaseExperimental && !experimental {
-				continue
-			}
+	return filter, nil
+}
+
+func (filter categoriesFilter) Filter(ctx context.Context, packages util.Packages) map[string]util.Package {
+	span, _ := apm.StartSpan(ctx, "FilterPackages", "custom")
```

nit: I assume it's `error` swallowed. Is it safe here?